src/Control/Exception/Safe.hs

```diff
@@ -344,27 +387,56 @@ tryAnyDeep = tryDeep
 -- <https://github.com/fpco/safe-exceptions#quickstart>
 --
 -- @since 0.1.0.0
-tryAsync :: (C.MonadCatch m, E.Exception e) => m a -> m (Either e a)
+tryAsync :: HAS_CALL_STACK => (C.MonadCatch m, E.Exception e) => m a -> m (Either e a)
 tryAsync = C.try
 
 -- | A variant of 'try' that takes an exception predicate to select
 -- which exceptions are caught.
 --
 -- @since 0.1.4.0
-tryJust :: (C.MonadCatch m, Exception e) => (e -> Maybe b) -> m a -> m (Either b a)
-tryJust f a = catch (Right `liftM` a) (\e -> maybe (throwM e) (return . Left) (f e))
+tryJust :: HAS_CALL_STACK => (C.MonadCatch m, Exception e) => (e -> Maybe b) -> m a -> m (Either b a)
+tryJust f a = withFrozenCallStack catch (Right `liftM` a) (\e -> maybe (throwM e) (return . Left) (f e))
 
 -- | Async safe version of 'E.onException'
 --
 -- @since 0.1.0.0
-onException :: C.MonadMask m => m a -> m b -> m a
-onException thing after = withException thing (\(_ :: SomeException) -> after)
+onException :: HAS_CALL_STACK => C.MonadMask m => m a -> m b -> m a
+onException thing after = withFrozenCallStack withException thing (\(_ :: SomeException) -> after)
+
+-- Note: [withFrozenCallStack impredicativity]
```

Comment:
I'm curious if `DeepSubsumption` helps here? https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0511-deep-subsumption.rst

Reply:
@parsonsmatt Good idea! Unfortunately, I just tried it and received the same error:

```haskell
{-# LANGUAGE DeepSubsumption #-}
...
withException thing after = withFrozenCallStack C.uninterruptibleMask $ ...

-- src/Control/Exception/Safe.hs:441:49: error: [GHC-83865]
--     • Couldn't match type: m a -> m a
--                      with: forall a1. m a1 -> m a1
--       Expected: (forall a1. m a1 -> m a1) -> m a
--         Actual: (m a -> m a) -> m a
--     • In the first argument of ‘withFrozenCallStack’, namely
--         ‘C.uninterruptibleMask’
```

This however inspired me to try manually eta-expanding All The Things, and it actually worked without any new extensions!

```diff
diff --git a/src/Control/Exception/Safe.hs b/src/Control/Exception/Safe.hs
index 0c4b5e9..8872d3c 100644
--- a/src/Control/Exception/Safe.hs
+++ b/src/Control/Exception/Safe.hs
@@ -437,7 +437,7 @@ onException thing after = withFrozenCallStack withException thing (\(_ :: SomeEx
 --
 -- @since 0.1.0.0
 withException :: HAS_CALL_STACK => (C.MonadMask m, E.Exception e) => m a -> (e -> m b) -> m a
-withException thing after = C.uninterruptibleMask $ \restore -> do
+withException thing after = withFrozenCallStack (\io -> C.uninterruptibleMask (\r -> io r)) $ \restore -> do
     fmap fst $ C.generalBracket (pure ()) cAfter (const $ restore thing)
   where
     -- ignore the exception from after, see bracket for explanation
@@ -464,7 +464,7 @@ bracket_ before after thing = withFrozenCallStack bracket before (const after) (
 --
 -- @since 0.1.0.0
 finally :: HAS_CALL_STACK => C.MonadMask m => m a -> m b -> m a
-finally thing after = C.uninterruptibleMask $ \restore -> do
+finally thing after = withFrozenCallStack (\io -> C.uninterruptibleMask (\r -> io r)) $ \restore -> do
     fmap fst $ C.generalBracket (pure ()) cAfter (const $ restore thing)
   where
     -- ignore the exception from after, see bracket for explanation
```

I missed this when I was experimenting earlier. I am curious what others think of this idea.